pkg/scannerv4/mappers/mappers.go

```diff
@@ -671,137 +643,86 @@ func pkgFixedBy(enrichments map[string][]json.RawMessage) (map[string]string, er
-func severityAndCvssMetrics(ctx context.Context, vuln *claircore.Vulnerability, nvdVuln *nvdschema.CVEAPIJSON20CVEItem) ([]*v4.VulnerabilityReport_Vulnerability_CVSS, *v4.VulnerabilityReport_Vulnerability_CVSS, string, error) {
+// cvssMetrics processes the CVSS metrics and severity for a given vulnerability.
+// This function gathers CVSS metrics data from multiple sources and
+// returns a slice of CVSS metrics collected from different sources (e.g., RHEL, NVD, OSV).
+// When not empty, the first entry is the "preferred" metric.
+// An error is returned when there is a failure to collect CVSS metrics from all sources;
+// however, the returned slice of metrics will still be populated with any successfully gathered metrics.
+// It is up to the caller to ensure the returned slice is populated prior to using it.
+func cvssMetrics(ctx context.Context, vuln *claircore.Vulnerability, nvdVuln *nvdschema.CVEAPIJSON20CVEItem) ([]*v4.VulnerabilityReport_Vulnerability_CVSS, error) {
 	// Add context values for consistent logging
 	ctx = zlog.ContextWithValues(ctx,
-		"component", "mappers/severityAndCvssMetrics",
+		"component", "mappers/cvssMetrics",
 		"updater", vuln.Updater,
 		"vuln_id", vuln.ID,
 		"vuln_name", vuln.Name,
 	)
 
-	var cvssMetrics []*v4.VulnerabilityReport_Vulnerability_CVSS
-
-	// RHEL CVSS metrics
-	// ignore errors since invalid sev values will trigger error in toCVSSMetric
-	rhelSev, _ := rhelSeverityAndScores(vuln)
-	rhelCVSS, rhelErr := toCVSSMetric(rhelSev)
-	if rhelErr == nil {
-		cvssMetrics = append(cvssMetrics, rhelCVSS)
-	}
+	var metrics []*v4.VulnerabilityReport_Vulnerability_CVSS
 
-	// NVD CVSS metrics
-	nvdSev, _ := nvdSeverityAndScores(vuln, nvdVuln)
-	nvdCVSS, nvdErr := toCVSSMetric(nvdSev)
-	if nvdErr == nil {
-		cvssMetrics = append(cvssMetrics, nvdCVSS)
-	}
-
-	// Undecided CVSS metrics
-	// ignore error since invalid sev values will trigger error in toCVSSMetric
-	undecidedSev, _ := cvssVector(vuln.Severity)
-	var undecidedCVSS *v4.VulnerabilityReport_Vulnerability_CVSS
-	var undecidedErr error
-	undecidedSev.url = vuln.Links
+	var preferredCVSS *v4.VulnerabilityReport_Vulnerability_CVSS
+	var preferredErr error
 	switch {
-	case strings.HasPrefix(vuln.Updater, osvUpdaterPrefix):
-		undecidedSev.source = v4.VulnerabilityReport_Vulnerability_CVSS_SOURCE_OSV
-		undecidedCVSS, undecidedErr = toCVSSMetric(undecidedSev)
+	case rhelUpdaterPattern.MatchString(vuln.Updater):
+		preferredCVSS, preferredErr = rhelCVSS(vuln)
+	case strings.HasPrefix(vuln.Updater, osvUpdaterPrefix) && !isOSVDBSpecificSeverity(vuln.Severity):
+		preferredCVSS, preferredErr = vulnCVSS(vuln, v4.VulnerabilityReport_Vulnerability_CVSS_SOURCE_OSV)
 	case strings.EqualFold(vuln.Updater, constants.ManualUpdaterName):
-		undecidedSev.source = v4.VulnerabilityReport_Vulnerability_CVSS_SOURCE_UNKNOWN
-		undecidedCVSS, undecidedErr = toCVSSMetric(undecidedSev)
+		preferredCVSS, preferredErr = vulnCVSS(vuln, sourceFromLinks(vuln.Links))
```

Comment:
I'm not sure if it's a bit redundant, but we could add a comment saying that vulnerabilities from the manual updater always have only one URL in `Links`. I've noticed that other data sources often provide multiple URLs.